### PR TITLE
CSP PDF uploads

### PR DESCRIPTION
**Pull request by richard-dds**
Upload TO pdfs using either AWS or Azure. Makes use of either CSP's methods of pre-authentication so that we can upload securely from the frontend without exposing secrets.

This PR does not include a way for us to download the files from the CSP, but it does provide enough of a base to make that a relatively simple future change.

### CODE REVIEW

File: atst/domain/csp/file_uploads.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            self.config["AWS_BUCKET_NAME"],
+            object_name,
+            ExpiresIn=3600,
+            Conditions=[
+                ("eq", "$Content-Type", "application/pdf"),
+                ("starts-with", "$x-amz-meta-filename", ""),
+            ],
+            Fields={"Content-Type": "application/pdf", "x-amz-meta-filename": ""},
+        )
+        return (presigned_post, object_name)
```
Comment (dandds) [inline atst/domain/csp/file_uploads.py]: I think both `get_token` methods could use a docstring explaining what they return, particularly the first part of the tuple. I had a little trouble following why the Azure version returns a dictionary with `{"token": {"token": ...`, and the only way to infer what the AWS version returns is to look at what properties the JS uploader class tries to access.

**Comment by dandds** [comment]
I also wonder whether it's better to use CSP-specific config value names (`AWS_`, `AZURE_`) or if the structure of the creds has enough in common we can use generic config value names (`CSP_STORAGE_SECRET_KEY`).

**Comment by richard-dds** [comment]
> I also wonder whether it's better to use CSP-specific config value names (`AWS_`, `AZURE_`) or if the structure of the creds has enough in common we can use generic config value names (`CSP_STORAGE_SECRET_KEY`).

@dandds I'm not sure if the config structures have enough in common for us to do this without making the key names confusing and asymmetrical.

File: js/lib/upload.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    form.set('x-amz-meta-filename', file.name)
+
+    return fetch(this.presignedPost.url, {
+      method: 'POST',
+      body: form,
+    })
+  }
+}
+
+export const buildUploader = token => new AzureUploader(token)
```
[the remark this answers is not in the document]
Comment (dandds) [reply js/lib/upload.js]: NVM, got it! Forgot to swap the INI config for the CSP.

File: js/components/upload_input.js [hunk 3]
```diff
@@ -52,9 +61,19 @@ export default {
...
+      try {
+        await this.uploader.upload(file, this.objectName)
+        this.attachment = e.target.value
+        this.showErrors = false
+        this.$refs.attachmentFilename.value = file.name
+        this.$refs.attachmentObjectName.value = this.objectName
+      } catch (err) {
+        console.log(err)
+        this.showErrors = true
+      }
```
[the remark this answers is not in the document]
Comment (richard-dds) [reply js/components/upload_input.js]: Added a story: https://www.pivotaltracker.com/story/show/167717072

File: templates/components/upload_input.html [hunk 4]
```diff
@@ -41,9 +43,11 @@
           aria-label="Task Order Upload"
           v-bind:value="attachment"
           type="file">
+          <input type="hidden" name="{{ field.filename.name }}" id="{{ field.filename.name }}" ref="attachmentFilename">
+          <input type="hidden" name="{{ field.object_name.name }}" id="{{ field.object_name.name }}" ref="attachmentObjectName">
       </div>
-      {% for error in field.errors %}
-        <span v-show="showErrors" class="usa-input__message">{{error}}</span>
+      {% for error, error_message in field.errors.items() %}
+        <span v-show="showErrors" class="usa-input__message">{{error_message}}</span>
```
Comment (tomdds) [inline templates/components/upload_input.html]: `error_message` seems to resolve to a list of html escaped strings (`[&#39;Test Error Message&#39;]` in `upload_input_error_template.html`), do we need to pass this through a filter or iterator of some kind, or do we want to defer that an error handling PR?

File: atst/domain/csp/file_uploads.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
+from azure.storage.common import CloudStorageAccount
+from azure.storage.blob import ContainerPermissions
+from datetime import datetime, timedelta
+from uuid import uuid4
+
+import boto3
```
[the remark this answers is not in the document]
Comment (tomdds) [reply atst/domain/csp/file_uploads.py]: Following up on this, we could import the csp specific deps in the respective constructors and bind them as properties for use in the methods?

[the remark this answers is not in the document]
Comment (richard-dds) [reply atst/domain/csp/file_uploads.py]: We can import them in the constructors, but I'm not totally sure how imports in functions actually work. I guess I figured that `import foo` was global and idempotent

File: js/lib/upload.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+  async upload(file, objectName) {
+    const blobService = Azure.createBlobServiceWithSas(
+      'https://atat.blob.core.windows.net',
+      this.sasToken
+    )
+    const fileReader = new FileReader()
+    const options = {
+      contentSettings: {
+        contentType: 'application/pdf',
```
[the remark this answers is not in the document]
Comment (richard-dds) [reply js/lib/upload.js]: Yes they do

File: Pipfile [hunk 7]
```diff
@@ -23,6 +23,10 @@ lockfile = "*"
 "flask-rq2" = "*"
 werkzeug = "*"
 PyYAML = "*"
+azure-storage = "*"
+azure-storage-common = "*"
+azure-storage-file = "*"
```
[the remark this answers is not in the document]
Comment (richard-dds) [reply Pipfile]: Apparently not. I removed it, ran `pipenv clean`, and everything still works.

File: atst/domain/csp/file_uploads.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (richard-dds) [reply atst/domain/csp/file_uploads.py]: @tomdds Is this what you had in mind? https://github.com/dod-ccpo/atst/pull/994/commits/f1560d5d7f543c333be30aee32368be6a0d2f68c

File: js/lib/upload.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    this.token = token
+  }
+
+  async upload(file, objectName) {
+    return Promise.resolve({})
+  }
+}
+
+export const buildUploader = token => {
+  const cloudProvider = process.env.CLOUD_PROVIDER || 'mock'
```
[the remark this answers is not in the document]
Comment (richard-dds) [reply js/lib/upload.js]: @dandds @tomdds After doing a bit of testing and research, I think our best bet is to use a `.env` file. Parcel comes with `dotenv` bundled, so we can provide configuration in `.env` (or one of a few specific filenames that dotenv will search for) which is injected into `process.env` at build time.

https://parceljs.org/env.html

File: templates/components/upload_input.html (hunk 4 above)
Comment (richard-dds) [reply templates/components/upload_input.html]: @tomdds We decided to move this to another PR that will cover other validation and error-related things: https://www.pivotaltracker.com/story/show/167764682

**Pull request merged**